submarine-server/server-core/src/main/java/org/apache/submarine/server/environment/EnvironmentManager.java

```diff
@@ -0,0 +1,226 @@
+ * under the License.
+ */
+
+package org.apache.submarine.server.environment;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.ws.rs.core.Response.Status;
+
+import org.apache.ibatis.session.SqlSession;
+import org.apache.submarine.commons.utils.exception.SubmarineRuntimeException;
+import org.apache.submarine.server.api.environment.Environment;
+import org.apache.submarine.server.api.spec.EnvironmentSpec;
+import org.apache.submarine.server.database.utils.MyBatisUtil;
+import org.apache.submarine.server.environment.database.entity.EnvironmentEntity;
+import org.apache.submarine.server.environment.database.mappers.EnvironmentMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Environment Management
+ */
+public class EnvironmentManager {
+  
+  private static final Logger LOG =
+      LoggerFactory.getLogger(EnvironmentManager.class);
+
+  private static volatile EnvironmentManager manager;
+  
+  /**
+   * Get the singleton instance
+   * @return object
+   */
+  public static EnvironmentManager getInstance() {
+    if (manager == null) {
+      synchronized (EnvironmentManager.class) {
+        if (manager == null) {
+          manager = new EnvironmentManager();
+        }
+      }
+    }
+    return manager;
+  }
+
+  private EnvironmentManager() {
+    
+  }
+
+  /**
+   * Create Environment
+   * @param spec environment spec
+   * @return Environment environment
+   * @throws SubmarineRuntimeException the service error
+   * @throws MetaException 
+   */
+  public Environment createEnvironment(EnvironmentSpec spec)
+      throws SubmarineRuntimeException {
+    checkSpec(spec);
+    EnvironmentEntity entity = new EnvironmentEntity();
+    entity.setEnvironmentName(spec.getName());
+    entity.setEnvironmentSpec(
+        new GsonBuilder().disableHtmlEscaping().create().toJson(spec));
+    
+    LOG.info("add({})", entity.toString());
+    try (SqlSession sqlSession = MyBatisUtil.getMetastoreSqlSession()) {
+      EnvironmentMapper environmentMapper =
+          sqlSession.getMapper(EnvironmentMapper.class);
+      int environmentId = environmentMapper.insert(entity);
+      sqlSession.commit();
+
+      Environment env = new Environment();
+      env.setName(spec.getName());
+      env.setEnvironmentId(environmentId);
+      env.setEnvironmentSpec(spec);
+      return env;
+    } catch (Exception e) {
+      LOG.error(e.getMessage(), e);
+      throw new SubmarineRuntimeException(Status.BAD_REQUEST.getStatusCode(),
+          "Unable to process the environment spec.");
+    }
+  }
+  
+  /**
+   * Update environment
+   * @param name Name of the environment
+   * @param spec environment spec
+   * @return Environment environment
+   * @throws SubmarineRuntimeException the service error
+   */
+  public Environment updateEnvironment(String name, EnvironmentSpec spec)
```

Comment:
It looks like updateEnvironment and createEnvironment have duplicated logic, can you try to dedup?

Reply:
updateEnvironment and createEnvironment calls update and insert method of db mapper class respectively. Hence the different methods. (or) Am I missing anything?

Reply:
There's only one line difference between the method, it's better to consolidate the two methods. Since it is not a trivial method, dedup can help us reduce maintenance cost, thus I suggested to do that.

Reply:
Taken care.

Reply:
Thanks.